File: 1481-least-number-of-unique-integers-after-k-removals/1481-least-number-of-unique-integers-after-k-removals.java
class Solution {
    public int findLeastNumOfUniqueInts(int[] arr, int k) {
        Map<Integer, Integer> freq = new HashMap<>();

        for (int x : arr) {
            freq.put(x, freq.getOrDefault(x, 0) + 1);
        }

        List<Integer> counts = new ArrayList<>();
        for (int cnt : freq.values()) {
            counts.add(cnt);
        }

        Collections.sort(counts);

        int unique = counts.size();

        for (int f : counts) {
            if (f <= k) {
                k -= f;
                unique--;
            } else {
                break;
            }
        }
        return unique;
    }
}

File: 1481-least-number-of-unique-integers-after-k-removals/1481-least-number-of-unique-integers-after-k-removals.py
class Solution:
    def findLeastNumOfUniqueInts(self, arr: List[int], k: int) -> int:

        freq = defaultdict(int)

        for x in arr:
            freq[x] += 1
        
        print(freq)

        freq_sorted = sorted(freq.items(),  key = lambda x: x[1])
        print(freq_sorted)
        unique = len(freq_sorted)
        for val, frq in freq_sorted:
            if frq <= k:
                k -= frq
                unique -= 1
        
        return unique

        